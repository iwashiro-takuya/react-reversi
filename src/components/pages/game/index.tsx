import React from 'react';
import styled from 'styled-components';
import Board from 'src/components/organisms/board';
import Score from 'src/components/organisms/score';
import Const from 'src/const';
import SettingForm from 'src/components/organisms/setting-form';
import useGame from './hooks';

const { Color } = Const;

export interface GameProps {
  dataCy: string;
}
const Game: React.FC<GameProps> = ({ dataCy }) => {
  const { isGameStarted } = useGame();

  return (
    <>
      <StyledHeading>React Reversi</StyledHeading>
      <StyledGameWrapper data-cy={dataCy}>
        {isGameStarted ? (
          <>
            <Board dataCy="board" />
            <Score />
          </>
        ) : (
          <SettingForm />
        )}
      </StyledGameWrapper>
    </>
  );
};

const StyledHeading = styled.h1`
  margin: 12px 0;
  text-align: center;
`;

const StyledGameWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 1026px;
  height: 768px;
  margin: 0 auto;
  color: ${Color.TX_BLACK};
`;

export default Game;
