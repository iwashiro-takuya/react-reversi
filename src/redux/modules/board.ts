import { BoardStateType } from 'src/@types';

const SET_BOARD = 'reversi/board/SET_BOARD';

type SetBoard = {
  type: typeof SET_BOARD;
  payload: BoardStateType;
};

type ActionType = SetBoard;

const initialState = {
  board: [],
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export default (state = initialState, action: ActionType) => {
  switch (action.type) {
    case SET_BOARD:
      return {
        ...state,
        board: action.payload,
      };
    default:
      return state;
  }
};

// updating...
