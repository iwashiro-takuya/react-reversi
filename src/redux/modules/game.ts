import { Dispatch } from 'redux';
import Const from 'src/const';

const { Player, PieceColor } = Const;

/**
 * action
 */
const UPDATE_GAME_START_FLAG = 'reversi/store/game/update-game-start-flag';
const UPDATE_SIDE_SQUARES_COUNT =
  'reversi/store/game/update-side-squares-count';
const UPDATE_BOARD = 'reversi/store/game/update-board';
const UPDATE_PLAYERS = 'reversi/store/game/update-players';

/**
 * action types
 */
type UpdateGameStartFlag = {
  type: typeof UPDATE_GAME_START_FLAG;
  payload: boolean;
};

type UpdateSideSquaresCount = {
  type: typeof UPDATE_SIDE_SQUARES_COUNT;
  payload: number;
};

type UpdateBoard = {
  type: typeof UPDATE_BOARD;
  payload: Board;
};

type UpdatePlayers = {
  type: typeof UPDATE_PLAYERS;
  payload: Store['game']['players'];
};

type ActionType =
  | UpdateGameStartFlag
  | UpdateSideSquaresCount
  | UpdateBoard
  | UpdatePlayers;

/**
 * initial state
 */
export const initialState: Store['game'] = {
  isGameStarted: false,
  sideSquaresCount: 0,
  board: [],
  players: {
    black: {
      player: { ...Player.PLAYER_1 },
      pieceColor: PieceColor.BLACK,
      score: 0,
      current: true,
    },
    white: {
      player: { ...Player.PLAYER_2 },
      pieceColor: PieceColor.WHITE,
      score: 0,
      current: false,
    },
  },
};

/**
 * reducer
 */
export default (
  state: Store['game'] = initialState,
  action: ActionType,
): Store['game'] => {
  switch (action.type) {
    case UPDATE_GAME_START_FLAG:
      return {
        ...state,
        isGameStarted: action.payload,
      };
    case UPDATE_SIDE_SQUARES_COUNT:
      return {
        ...state,
        sideSquaresCount: action.payload,
      };
    case UPDATE_BOARD:
      return {
        ...state,
        board: action.payload,
      };
    case UPDATE_PLAYERS:
      return {
        ...state,
        players: action.payload,
      };
    default:
      return state;
  }
};

export const updateGameStartFlg = (flag: boolean): UpdateGameStartFlag => ({
  type: UPDATE_GAME_START_FLAG,
  payload: flag,
});

export const updateSideSquaresCount = (
  sideSquaresCount: number,
): UpdateSideSquaresCount => ({
  type: UPDATE_SIDE_SQUARES_COUNT,
  payload: sideSquaresCount,
});

export const updateBoard = (stagingBoard: Board): UpdateBoard => ({
  type: UPDATE_BOARD,
  payload: stagingBoard,
});

export const updatePlayers = (
  stagingPlayers: Store['game']['players'],
): UpdatePlayers => ({
  type: UPDATE_PLAYERS,
  payload: stagingPlayers,
});

export const initializeBoard = (sideSquaresCount: number) => (
  dispatch: Dispatch,
): void => {
  const squaresCountAmount: number = sideSquaresCount ** 2;
  const stagingBoard: Board = [];

  for (
    let squareCount = 0;
    squareCount < squaresCountAmount;
    squareCount += 1
  ) {
    const key = squareCount;
    const column = squareCount % sideSquaresCount;
    const row = Math.floor(squareCount / sideSquaresCount);
    let pieceColor: UnionVal<typeof PieceColor> = PieceColor.INVISIBLE;

    /**
     * place four stones when the board is initially rendered
     */
    // upprer left square
    if (column === sideSquaresCount / 2 && row === sideSquaresCount / 2 - 1) {
      pieceColor = PieceColor.WHITE;
    }

    // upper right square
    if (
      column === sideSquaresCount / 2 - 1 &&
      row === sideSquaresCount / 2 - 1
    ) {
      pieceColor = PieceColor.BLACK;
    }

    // lower left square
    if (column === sideSquaresCount / 2 && row === sideSquaresCount / 2) {
      pieceColor = PieceColor.BLACK;
    }

    // lower right square
    if (column === sideSquaresCount / 2 - 1 && row === sideSquaresCount / 2) {
      pieceColor = PieceColor.WHITE;
    }

    stagingBoard.push({
      key,
      column,
      row,
      pieceColor,
    });
  }

  dispatch(updateSideSquaresCount(sideSquaresCount));
  dispatch(updateBoard(stagingBoard));
};

export const changeGamesTurn = (
  square: Square,
  updatableSquaresArray: Square[],
) => (dispatch: Dispatch, getState: () => Store): void => {
  const {
    game: { board: stagingBoard, players },
  } = getState();
  const clickedSquare: Square = stagingBoard[square.key];

  const currentPlayerIndex = Object.entries(players).find(
    ([_id, player]) => player.current === true,
  )?.[0] as keyof Store['game']['players'];
  const currentPlayer = players[currentPlayerIndex];

  // change each the value of squares for pieces to be turn over
  updatableSquaresArray.forEach((updatableSquare: Square) => {
    stagingBoard[updatableSquare.key].pieceColor = currentPlayer.pieceColor;
  });

  // change the value of clicked square
  clickedSquare.pieceColor = currentPlayer.pieceColor;

  // switch the current player
  const stagingPlayers = Object.fromEntries(
    Object.entries(players).map(([key, player]) => {
      const stagingPlayer = { ...player };
      stagingPlayer.current = !stagingPlayer.current;

      return [key, stagingPlayer];
    }),
  ) as Store['game']['players'];

  dispatch(updateBoard(stagingBoard));
  dispatch(updatePlayers(stagingPlayers));
  updateScore();
};

export const updateScore = () => (
  dispatch: Dispatch,
  getState: () => Store,
): void => {
  const {
    game: { board, players },
  } = getState();

  const whitePiecesCount = board.reduce((prev, square) => {
    return prev + (square.pieceColor === PieceColor.WHITE ? 1 : 0);
  }, 0);

  const blackPiecesCount = board.reduce((prev, square) => {
    return prev + (square.pieceColor === PieceColor.BLACK ? 1 : 0);
  }, 0);

  const stagingPlayers = { ...players };

  stagingPlayers.white.score = whitePiecesCount;
  stagingPlayers.black.score = blackPiecesCount;

  dispatch(updatePlayers(stagingPlayers));
};
