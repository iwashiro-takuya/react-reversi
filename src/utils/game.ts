import PieceColor from 'src/const/piece-color';
import store from 'src/redux/create-store';

/**
 * Function to get arrays of the square that has piece that can be turn over in each directions from the passed square
 */
const getUpdatableSquaresArray = (passedSquare: Square): Square[] => {
  const { board, sideSquaresCount, players } = store.getState().game;

  const currentPlayerIndex = Object.entries(players).find(
    ([_id, player]) => player.current === true,
  )?.[0] as UnionVal<PieceColorType>;
  const currentPlayer = players[currentPlayerIndex];

  const eachDirectionSquaresArray: Array<Square[]> = [];

  // Get an array of squares from the passed square to the left
  eachDirectionSquaresArray.push(
    board
      .filter((square) => {
        return !!(
          square.row === passedSquare.row && square.column < passedSquare.column
        );
      })
      .reverse(), // Reverse to use the passed square as a starting point
  );

  // Get an array of squares from the passed square to the right side
  eachDirectionSquaresArray.push(
    board.filter((square) => {
      return !!(
        square.row === passedSquare.row && square.column > passedSquare.column
      );
    }),
  );

  // Get an array of squares from the passed square to the top
  eachDirectionSquaresArray.push(
    board
      .filter((square) => {
        return !!(
          square.row < passedSquare.row && square.column === passedSquare.column
        );
      })
      .reverse(), // Reverse to use the passed square as a starting point
  );

  // Get an array of squares from the passed square to the bottom
  eachDirectionSquaresArray.push(
    board.filter((square) => {
      return !!(
        square.row > passedSquare.row && square.column === passedSquare.column
      );
    }),
  );

  // Get an array of squares from the passed square to the upper left
  const upperLeftSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key - sideSquaresCount * count - count;
    if (squareId >= 0 && squareId < board.length) {
      upperLeftSquaresArray.push(board[squareId]);
    }
  }
  eachDirectionSquaresArray.push(upperLeftSquaresArray);

  // Get an array of squares from the passed square to the upper right
  const upperRightSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key - sideSquaresCount * count + count;
    if (squareId > 0 && squareId < board.length) {
      upperRightSquaresArray.push(board[squareId]);
    }
  }
  eachDirectionSquaresArray.push(upperRightSquaresArray);

  // Get an array of squares from the passed square to the lower left
  const lowerLeftSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key + sideSquaresCount * count - count;
    if (squareId >= 0 && squareId < board.length - 1) {
      lowerLeftSquaresArray.push(board[squareId]);
    }
  }
  eachDirectionSquaresArray.push(lowerLeftSquaresArray);

  // Get an array of squares from the passed square to the lower right
  const lowerRightSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key + sideSquaresCount * count + count;
    if (squareId >= 0 && squareId < board.length) {
      lowerRightSquaresArray.push(board[squareId]);
    }
  }
  eachDirectionSquaresArray.push(lowerRightSquaresArray);

  const updatableSquaresArray: Square[] = [];

  eachDirectionSquaresArray.forEach((squareArray) => {
    const emptySquareIndex: number = squareArray.findIndex(
      (square) => square.pieceColor === PieceColor.INVISIBLE,
    );
    if (emptySquareIndex !== -1) {
      squareArray.splice(emptySquareIndex, squareArray.length);
    }

    const endpointSquareIndex: number = squareArray.findIndex(
      (square) => square.pieceColor === currentPlayer.pieceColor,
    );
    squareArray.splice(Math.max(0, endpointSquareIndex), squareArray.length);

    updatableSquaresArray.push(...squareArray);
  });

  return updatableSquaresArray;
};

export default { getUpdatableSquaresArray };
