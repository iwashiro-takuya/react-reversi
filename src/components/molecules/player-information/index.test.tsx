import React from 'react';
import renderer from 'react-test-renderer';
import { initialState } from 'src/redux/modules/game';
import PlayerInformation from '.';

describe('PlayerInformation unit test', () => {
  it('snapshot(PieceColor White)', () => {
    const component = renderer.create(
      <PlayerInformation gamePlayer={initialState.players.black} />,
    );
    expect(component).toMatchSnapshot();
  });
  it('snapshot(PieceColor Black)', () => {
    const component = renderer.create(
      <PlayerInformation gamePlayer={initialState.players.white} />,
    );
    expect(component).toMatchSnapshot();
  });
});
