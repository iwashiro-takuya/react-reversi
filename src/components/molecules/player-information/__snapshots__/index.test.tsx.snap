// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerInformation unit test snapshot(PieceColor Black) 1`] = `
<div
  className="sc-gtsrHT ljPrYI"
>
  <span
    className="sc-jSFjdj bsAtwI"
  >
    0
  </span>
  <div
    className="sc-dlnjwi gFWwNh"
  >
    <span
      className="sc-bdnxRM bGNMMU"
      color="#fff"
    />
    <p
      className="sc-eCApnc btUvBL"
      color="#555"
    >
      P2
    </p>
  </div>
</div>
`;

exports[`PlayerInformation unit test snapshot(PieceColor White) 1`] = `
<div
  className="sc-gtsrHT ljPrYI"
>
  <div
    className="sc-dlnjwi gFWwNh"
  >
    <span
      className="sc-bdnxRM kIMDaa"
      color="#000"
    />
    <p
      className="sc-eCApnc jWhorU"
      color="#fff"
    >
      P1
    </p>
  </div>
  <span
    className="sc-jSFjdj bsAtwI"
  >
    0
  </span>
</div>
`;
